### DISCUSSION TITLE
Is there a way to package C++ distributable as MSIX package

### DISCUSSION THREAD

**Discussion by harvinders** [microsoft/WindowsAppSDK#3646] [Q&A]
We have a WinAppSDK 1.3 based msix package for our application. Our application uses some C++ library that require C++ redistributables to be installed on the OS for it to work. I was searching for MSIX packages for the same but could not find them. 

I remember when WinAppSDK had two project solution, there were C++ UWP redistributable MSIX packages as dependencies. 

Is there any place I can find the equivalent MSIX packages for C++ redistributables (https://learn.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022) or is there a way to create a MSIX package out of those redistributables.

**Comment by DarranRowe** [comment]
The Visual C++ Runtime versions 11, 12 and 14 have packages. You can find documentation about this [here](https://learn.microsoft.com/en-us/troubleshoot/developer/visualstudio/cpp/libraries/c-runtime-packages-desktop-bridge#how-to-install-and-update-desktop-framework-packages). Don't worry about the .appx extension, these are what you want. The extension can be thought of as the package can install on older versions of Windows 10.
However, remember that you will need the package reference to use it.

**Reply by harvinders** [reply]
@DarranRowe I am looking for the latest C++ redistributables and the one mentioned [here](https://learn.microsoft.com/en-us/troubleshoot/developer/visualstudio/cpp/libraries/c-runtime-packages-desktop-bridge#how-to-install-and-update-desktop-framework-packages) does not work for me, in the sense it installs, but these are not what I need. I need the one mentioned [here](https://learn.microsoft.com/en-US/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022)

**Reply by DarranRowe** [reply]
What is it about those packages that "does not work"?
If you look at the content of the package (it is a standard .zip file, but with a different extension), you will see that it has the desktop VC runtime naming of the files.

![Screenshot 2023-05-29 010328](https://github.com/microsoft/WindowsAppSDK/assets/52577874/9283b207-2a5a-4e08-be52-452466992afb)

These are also the runtime libraries that the WinUI 3 project system uses to satisfy the dependencies when it generates a package.
```
  <Dependencies>
    <TargetDeviceFamily Name="Windows.Universal" MinVersion="10.0.17763.0" MaxVersionTested="10.0.22621.0" />
    <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.17763.0" MaxVersionTested="10.0.22621.0" />
    <PackageDependency Name="Microsoft.WindowsAppRuntime.1.3" MinVersion="3000.851.1712.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
    <PackageDependency Name="Microsoft.VCLibs.140.00" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
    <PackageDependency Name="Microsoft.VCLibs.140.00.UWPDesktop" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
  </Dependencies>
```
As an example. In the output shown in Visual Studio, the following are clearly visible.
```
'Meh.exe' (Win32): Loaded 'C:\Program Files\WindowsApps\Microsoft.VCLibs.140.00.UWPDesktop_14.0.30704.0_x64__8wekyb3d8bbwe\msvcp140.dll'. Symbols loaded.
'Meh.exe' (Win32): Loaded 'C:\Program Files\WindowsApps\Microsoft.VCLibs.140.00.UWPDesktop_14.0.30704.0_x64__8wekyb3d8bbwe\vcruntime140_1.dll'. Symbols loaded.
'Meh.exe' (Win32): Loaded 'C:\Program Files\WindowsApps\Microsoft.VCLibs.140.00.UWPDesktop_14.0.30704.0_x64__8wekyb3d8bbwe\vcruntime140.dll'. Symbols loaded.
```
So they certainly work and they do satisfy dependencies for C++. Of course, since the Windows App SDK and WinUI 3 is completely desktop now, any C++ reference would require the desktop version of the packages anyway.
So what is the problem? Do you need the exact file versions found in the 17.6 version of the redistributable? Are there unresolved libraries? Is it something else?

**Comment by ghost** [comment]
I'm having the same problem here :/
vcruntime140.dll is bundled in `executable_name.crt` folder
and works unpacked
but when packed as .misx with MakeAppX it complains about missing VCRUNTIME140.dll :/
pls help :/

**Reply by DarranRowe** [reply]
Considering you wrote:

>vcruntime140.dll is bundled in `executable_name.crt` folder

I don't think this is the same situation. While you would need to provide more information, this implies to me that you are using some form of DLL redirection. Have you bundled up the VC Redistributable as a private assembly?

Please provide more information on what you are doing. Also, since you stated that it is bundled in a subdirectory, do you really need that specific version of the VC Redistributable or are you bundling it for a different reason?

Anyway, remember that the [DLL search order for packaged applications](https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-packaged-apps) is much more restricted. There are also two more things to remember with packages. First, the working directory is normally %SystemRoot%\System32. (Normally C:\Windows\System32, or C:\Windows\SysWoW64 for 32 bit applications on x64 versions of Windows.*) Secondly, [you have another way to attempt to redirect these files](https://learn.microsoft.com/en-us/windows/msix/desktop/desktop-to-uwp-behind-the-scenes#common-file-system-operations).

*I'm not mentioning ARM64 versions of Windows here, that just adds confusion.

**Reply by ghost** [reply]
Ok many thanks I'll look at these,
basically what I'm trying to do is package https://github.com/Bforartists/Bforartists as a .msix
it is a Blender fork
and by default the CMake configuration outputs a .crt folder next to the entry point binary and it works unpacked, and even works packed on my dev machine (because my dev machine has VC runtime installed I suppose)

but the Windows Store review didn't pass, they say it fails to launch and displays VCRuntime140.dll missing error message. and I confirmed on a local virtual machine

I'm basically using cpack to create a ZIP package then unpacking that to Content/Bforartists
and using the AppManifest.xml.template to create a .msix

I copy AppManifest.xml.template from the repo (from release\windows\msix\AppxManifest.xml.template)
next to this Python file which generates the AppManifest.xml and creates the msix package, and optionally sign it

![image](https://github.com/microsoft/WindowsAppSDK/assets/25552173/e003a259-19d8-4724-a2a2-a69991d1a272)


```python
# Based on https://archive.blender.org/developer/D8310

import subprocess as sp
from pathlib import Path
import sys


if len(sys.argv) < 4:
    print("Usage: create_msix.py <command> <version> <type>")
    print("command: sign, build, build_and_sign")
    print("version: version number, example 4.1.1.0")
    print('type: "Local" or "Store"')
    sys.exit(1)

command = sys.argv[1]
version = sys.argv[2]
package_type = sys.argv[3]


if package_type == "Local":
    # https://community.qlik.com/t5/Official-Support-Articles/How-to-find-certificates-by-thumbprint-or-name-with-powershell/ta-p/1711332
    # https://techcommunity.microsoft.com/t5/msix/msix-packageing-tool-signtool-certificate-issues/m-p/2133583/highlight/true#M78
    # https://stackoverflow.com/a/3961530/8094047
    certificate_subject_line = (
        'CN="Open Source Developer, Eyad Ahmed", O=Open Source Developer, L=Alexandria, S=Alexandria, C=EG'
    )
    publisher = certificate_subject_line.replace('"', "&quot;")
elif package_type == "Store":
    # Windows store
    publisher = "CN=1A48DC4A-D493-4696-B683-9EF3353D30D6"
else:
    print('Invalid package type, must be either "Local" or "Store".')
    sys.exit(1)


# BFA does not maintain LTS versions
##
LTSORNOT = ""
PACKAGETYPE = ""
##

content_folder = Path(__file__).parent / "Content"


manifest_text = (Path(__file__).parent / "AppxManifest.xml.template").read_text()
manifest_text = manifest_text.replace("[VERSION]", version)
manifest_text = manifest_text.replace("[PUBLISHER]", publisher)
manifest_text = manifest_text.replace("[LTSORNOT]", LTSORNOT)
manifest_text = manifest_text.replace("[PACKAGETYPE]", PACKAGETYPE)
(content_folder / "AppxManifest.xml").write_text(manifest_text)

output_msix_name = f"Bforartists_{version}.msix"


msix_command = [
    "makeappx",
    "pack",
    "/h",
    "SHA256",
    "/d",
    f"{content_folder.absolute()}",
    "/p",
    output_msix_name,
]


signtool_command = [
    "signtool",
    "sign",
    "/sha1",
    "4ae3d89dc7440d4533b99fb4977e114e4a0f052d",
    "/tr",
    "http://time.certum.pl/",
    "/td",
    "sha256",
    "/fd",
    "sha256",
    "/v",
    output_msix_name,
]


def run_command(command):
    sp.run(command, stdin=sys.stdin, stderr=sys.stderr, stdout=sys.stdout, encoding="UTF-8")


if command == "build_and_sign":
    run_command(msix_command)
    run_command(signtool_command)
elif command == "build":
    run_command(msix_command)
elif command == "sign":
    run_command(signtool_command)
else:
    print('Invalid command, must be one of "sign", "build", "build_and_sign".')
    sys.exit(1)
```

**Reply by DarranRowe** [reply]
Well, just to let you know, that template does not have the VC Redistributable listed as a dependency. It needs to have:
```
<PackageDependency Name="Microsoft.VCLibs.140.00.UWPDesktop" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
```
In the Dependencies element. Since I can't be entirely sure that you are not doing anything extra to the template compared to what is shown, does the output manifest have:

```
  <Dependencies>
    <TargetDeviceFamily Name="Windows.Desktop" MinVersion="10.0.17763.0" MaxVersionTested="10.0.18335.0" />
    <PackageDependency Name="Microsoft.VCLibs.140.00.UWPDesktop" MinVersion="14.0.30704.0" Publisher="CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US" />
  </Dependencies>
```
or similar for the dependencies?

**Reply by ghost** [reply]
I tried adding "Microsoft.VCLibs.140.00" before
it didn't help, I guess I will try "Microsoft.VCLibs.140.00.UWPDesktop" then

**Reply by DarranRowe** [reply]
Microsoft.VCLibs.140.00 wouldn't work. That is for UWP applications and the library base names have the _app suffix (vcruntime140_app.dll as an example).

![Screenshot 2024-05-21 220925](https://github.com/microsoft/WindowsAppSDK/assets/52577874/4af1d4cf-7337-4544-8a0f-bf09393ac504)

The libraries wouldn't satisfy applications or libraries built for desktop.

Microsoft.VCLibs.140.00.UWPDesktop is for desktop applications. The files all have the correct names to satisfy packaged desktop applications.

![Screenshot 2024-05-21 221251](https://github.com/microsoft/WindowsAppSDK/assets/52577874/4d966776-e6f9-4ab5-a4b5-8fff93028874)

**Comment by ghost** [comment]
Thanks, I expected it to be the opposite actually, since VC14.0 doesn't have uwp in the name, but
.UWPDesktop has,
I'll try tomorrow then, since it is past midnight here

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
